### ISSUE TITLE
agentType "openai-functions" continues the chain after signal.abort()

### ISSUE THREAD

**Issue by dejecj** [langchain-ai/langchainjs#1910]
I am working on an assistant using a number of tools. The use-case is that I want to handle certain tools differently, for example on some tools I want the model to summarize what it wants to do and ask for confirmation before doing it. So my thought was that I can use some callbacks before the tool is invoked and, based on some conditions, stop the chain. I am using an agent executor and I have passed a signal from an abortController. The callback I am using is handleAgentAction. What I am seeing is that even after signal.abort() the tool is invoked. I setup a sleep function to see if it was happening before or after but sure enough even 10 seconds after the abort command the tool runs. 

After some trial and error it seems that this only happens when using agentType: openai-functions. It does not happen with "zero-shot-react-description" | "chat-zero-shot-react-description" | "chat-conversational-react-description".

Here is a snippet of code so you can see what I have. Currently I stop on every action and request confirmation but that would change to the real use-case if I can get this working.

```javascript
 if(!memory[id]){
      memory[id] = new BufferWindowMemory({
          memoryKey: "chat_history",
          returnMessages: true,
          k: 5
      });
  }

  const executor = await initializeAgentExecutorWithOptions(tools, model, {
      agentType: "openai-functions",
      verbose: true,
      memory: memory[id],
      returnMessages: true,
      agentArgs: {
          prefix: `You are a friendly AI assistant. Answer the following questions truthfully and as best as you can. The UTC time is ${new Date().toString()}.`,
          memoryPrompts: [new MessagesPlaceholder("chat_history")],
          inputVariables: ["input", "agent_scratchpad", "chat_history"],
      },
      callbacks: [
          {
              handleAgentAction(action) {
                  controller.abort({
                      reason: 'need_confirmation',
                      action
                  });
              },
          }
      ]
  });

  const result = await executor.call({
      input: messages.at(-1).content,
      signal: controller.signal
  });
```

Basically in the action in handleAgentAction still fires. My understanding was that this hook is called "right before" the agent uses a tool as mentioned in the API reference. So I would think that cancelling the operation at this point would stop the function from being called. Might be an issue in the openai-functions agent implementation?

What do you guys think?

**Comment by jacoblee93** [comment]
Hey, thank you for opening this - will have a look and report back as soon as I can!

**Comment by dejecj** [comment]
Ok here I've prepared a test case and will share the code and the chain logs that show what I am experiencing:

This is the code I used for the demonstration:

```
"use server"
import { z } from "zod";
/**
 * START Langchain Imports
 */
import { ChatOpenAI } from "langchain/chat_models/openai";
import { initializeAgentExecutorWithOptions } from "langchain/agents";
import { DynamicStructuredTool } from "langchain/tools";
import { BufferWindowMemory } from "langchain/memory";
import { MessagesPlaceholder } from "langchain/prompts";
import { HumanChatMessage } from "langchain/schema";
/**
 * END Langchain Imports
 */
/** --- */
/**
 * START Langchain Tools
 */
const tools = [
    new DynamicStructuredTool({
        name: "list_accounts",
        description: "Retrieves a list or count of accounts based on specific search parameters provided in the query object.",
        schema: z.object({
            query: z.object({
                _id: z.string().default("").describe("Unique identifier of the account in the database."),
                active: z.void(z.boolean()).describe("Active status of the account. True for active accounts, false for archived accounts."),
                referred_by: z.string().default("").describe("Unique identifier of the affiliate account that referred this account."),
                email: z.any().default("").describe("Email address associated with the account."),
                name: z.string().default("").describe("Name associated with the account."),
                'owner.name': z.any().default("").describe("Name of the account owner."),
                main: z.enum(["main", "sub", "both"]).default('both').describe("Account type to filter the search. Can be 'main' for main accounts, 'sub' for sub accounts, or 'both' for both types."),
                created_at: z.any().default("").describe("Date range for when the account was created in UTC. Expected format: 'YYYY-MM-DDT00:00:00Z'. The function always expects a date range, even if it's the start and end of the same day."),
            }).required(),
            sort: z.string().default("").describe("Sort order for the returned results. For example, '-created_at' for descending order by creation date, '+created_at' for ascending order by creation date."),
            limit: z.any().describe("Maximum number of results to return. The maximum limit is capped at 500."),
            operation_type: z.enum(["list", "count"]).default('list').describe("Type of data that should be returned. 'list' will return a list of documents. 'count' will return the total number of documents."),
            select: z.string().default('').describe("Space-separated list of fields to be included in the returned results.")
        }),
        func: async () => '99'
    })
]
/**
 * END Langchain Tools
 */
/** --- */
/** 
 * START Langchain Configs 
 */
const model = new ChatOpenAI({
    temperature: 0.9,
    modelName: "gpt-3.5-turbo-16k",
});

/** 
 * END Langchain Configs 
 */
/** --- */

const memory = {}

export async function functionCompletion(id, messages) {
    const controller = new AbortController();
    try {
        if(!memory[id]){
            memory[id] = new BufferWindowMemory({
                memoryKey: "chat_history",
                returnMessages: true,
                k: 5
            });
        }

        const executor = await initializeAgentExecutorWithOptions(tools, model, {
            agentType: "openai-functions",
            verbose: true,
            memory: memory[id],
            returnMessages: true,
            agentArgs: {
                prefix: `You are Annie, a friendly AI assistant. Answer the following questions truthfully and as best you can. The UTC time is ${new Date().toString()}.`,
                memoryPrompts: [new MessagesPlaceholder("chat_history")],
                inputVariables: ["input", "agent_scratchpad", "chat_history"],
            },
            callbacks: [
                {
                    handleAgentAction(action) {
                        controller.abort({
                            reason: 'need_confirmation',
                            action
                        });
                    },
                }
            ]
        });

        const result = await executor.call({
            input: messages.at(-1).content,
            signal: controller.signal
        });

        return {
            messages: [...messages, { role: 'assistant', content: result.output}],
            responses: []
        }
    } catch(err){
        if (controller.signal?.aborted){
            const confirmation = await model.call([
                new HumanChatMessage(`Given this proposed action_payload, generate an explanation of the proposed action in simple terms and ask the user for confirmation: 
                
                action_payload: ${JSON.stringify(controller.signal.reason.action)}.`),
            ], {
                memory: memory[id]
            });
            return {
                messages: [...messages, { role: 'assistant', content: confirmation.text, function_call: {name:controller.signal.reason.action.tool, arguments: controller.signal.reason.action.toolInput}}],
                responses: []
            }
        }
        return [{ error: err.message }]
    }
}
```

For the sake of running the test, messages is an array of objects with a content prop.

When I run functionCompletion I get the following output:

```
[chain/start] [1:chain:AgentExecutor] Entering Chain run with input: {
  "input": "Introduce your self",
  "signal": {},
  "chat_history": []
}
[llm/start] [1:chain:AgentExecutor > 2:llm:ChatOpenAI] Entering LLM run with input: {
  "messages": [
    [
      {
        "type": "system",
        "data": {
          "content": "You are Annie, a friendly AI assistant. Answer the following questions truthfully and as best you can. The UTC time is Mon Jul 10 2023 21:22:56 GMT+0000 (Coordinated Universal Time).",
          "additional_kwargs": {}
        }
      },
      {
        "type": "human",
        "data": {
          "content": "Introduce your self",
          "additional_kwargs": {}
        }
      }
    ]
  ]
}
[llm/end] [1:chain:AgentExecutor > 2:llm:ChatOpenAI] [724ms] Exiting LLM run with output: {
  "generations": [
    [
      {
        "text": "Hello! I am Annie, a friendly AI assistant. I am here to assist you with any questions or tasks you may have. How can I help you today?",
        "message": {
          "type": "ai",
          "data": {
            "content": "Hello! I am Annie, a friendly AI assistant. I am here to assist you with any questions or tasks you may have. How can I help you today?",
            "additional_kwargs": {}
          }
        }
      }
    ]
  ],
  "llmOutput": {
    "tokenUsage": {
      "completionTokens": 34,
      "promptTokens": 286,
      "totalTokens": 320
    }
  }
}
[chain/end] [1:chain:AgentExecutor] [735ms] Exiting Chain run with output: {
  "output": "Hello! I am Annie, a friendly AI assistant. I am here to assist you with any questions or tasks you may have. How can I help you today?"
}
[chain/start] [1:chain:AgentExecutor] Entering Chain run with input: {
  "input": "How many accounts do we have?\n\n",
  "signal": {},
  "chat_history": [
    {
      "type": "human",
      "data": {
        "content": "Introduce your self",
        "additional_kwargs": {}
      }
    },
    {
      "type": "ai",
      "data": {
        "content": "Hello! I am Annie, a friendly AI assistant. I am here to assist you with any questions or tasks you may have. How can I help you today?",
        "additional_kwargs": {}
      }
    }
  ]
}
[llm/start] [1:chain:AgentExecutor > 2:llm:ChatOpenAI] Entering LLM run with input: {
  "messages": [
    [
      {
        "type": "system",
        "data": {
          "content": "You are Annie, a friendly AI assistant. Answer the following questions truthfully and as best you can. The UTC time is Mon Jul 10 2023 21:23:02 GMT+0000 (Coordinated Universal Time).",
          "additional_kwargs": {}
        }
      },
      {
        "type": "human",
        "data": {
          "content": "Introduce your self",
          "additional_kwargs": {}
        }
      },
      {
        "type": "ai",
        "data": {
          "content": "Hello! I am Annie, a friendly AI assistant. I am here to assist you with any questions or tasks you may have. How can I help you today?",
          "additional_kwargs": {}
        }
      },
      {
        "type": "human",
        "data": {
          "content": "How many accounts do we have?\n\n",
          "additional_kwargs": {}
        }
      }
    ]
  ]
}
[llm/end] [1:chain:AgentExecutor > 2:llm:ChatOpenAI] [497ms] Exiting LLM run with output: {
  "generations": [
    [
      {
        "text": "",
        "message": {
          "type": "ai",
          "data": {
            "content": "",
            "additional_kwargs": {
              "function_call": {
                "name": "list_accounts",
                "arguments": "{\n  \"query\": {}\n}"
              }
            }
          }
        }
      }
    ]
  ],
  "llmOutput": {
    "tokenUsage": {
      "completionTokens": 13,
      "promptTokens": 334,
      "totalTokens": 347
    }
  }
}
[agent/action] [1:chain:AgentExecutor] Agent selected action: {
  "tool": "list_accounts",
  "toolInput": {
    "query": {}
  },
  "log": ""
}
[chain/error] [1:chain:AgentExecutor] [504ms] Chain run errored with error: "AbortError"
[tool/start] [1:tool:DynamicStructuredTool] Entering Tool run with input: "{"query":{"_id":"","referred_by":"","email":"","name":"","owner.name":"","main":"both","created_at":""},"sort":"","operation_type":"list","select":""}"
[tool/end] [1:tool:DynamicStructuredTool] [1ms] Exiting Tool run with output: "99"
[llm/start] [1:llm:ChatOpenAI] Entering LLM run with input: {
  "messages": [
    [
      {
        "type": "system",
        "data": {
          "content": "You are Annie, a friendly AI assistant. Answer the following questions truthfully and as best you can. The UTC time is Mon Jul 10 2023 21:23:02 GMT+0000 (Coordinated Universal Time).",
          "additional_kwargs": {}
        }
      },
      {
        "type": "human",
        "data": {
          "content": "Introduce your self",
          "additional_kwargs": {}
        }
      },
      {
        "type": "ai",
        "data": {
          "content": "Hello! I am Annie, a friendly AI assistant. I am here to assist you with any questions or tasks you may have. How can I help you today?",
          "additional_kwargs": {}
        }
      },
      {
        "type": "human",
        "data": {
          "content": "How many accounts do we have?\n\n",
          "additional_kwargs": {}
        }
      },
      {
        "type": "ai",
        "data": {
          "content": "",
          "additional_kwargs": {
            "function_call": {
              "name": "list_accounts",
              "arguments": "{\"query\":{}}"
            }
          }
        }
      },
      {
        "type": "function",
        "data": {
          "content": "99",
          "name": "list_accounts",
          "additional_kwargs": {}
        }
      }
    ]
  ]
}
[llm/error] [1:llm:ChatOpenAI] [5ms] LLM run errored with error: "Cancel: canceled"
```

Looking towards the end of those logs you will see:
```
{
    "type": "function",
    "data": {
      "content": "99",
      "name": "list_accounts",
      "additional_kwargs": {}
    }
  }
```

Indicating the function was actually called.

Thanks again for the help.

**Comment by jacoblee93** [comment]
I see - looking at the code, there shouldn't actually be any difference in how different agent types are handled in the callback (the functions agent still uses an executor), and when I tried with another agent type, it seemed to exhibit the same behavior of calling the tool:

https://github.com/hwchase17/langchainjs/blob/main/langchain/src/agents/executor.ts#L119

Rather than using a signal there, you could try throwing an error or `await`'ing some long-running function that collects user input?

**Comment by dejecj** [comment]
@jacoblee93 So I did as suggested and instead of using a signal I instead throw an error from the callback. That seems to behave the in a similar way as the signal. See the logs below:

```
[llm/end] [1:chain:AgentExecutor > 2:llm:ChatOpenAI] [581ms] Exiting LLM run with output: {
  "generations": [
    [
      {
        "text": "",
        "message": {
          "type": "ai",
          "data": {
            "content": "",
            "additional_kwargs": {
              "function_call": {
                "name": "list_accounts",
                "arguments": "{\n  \"query\": {\n    \"main\": \"main\"\n  },\n  \"operation_type\": \"count\"\n}"
              }
            }
          }
        }
      }
    ]
  ],
  "llmOutput": {
    "tokenUsage": {
      "completionTokens": 30,
      "promptTokens": 342,
      "totalTokens": 372
    }
  }
}
Error in handler Handler, handleAgentAction: Error: need_confirmation
[agent/action] [1:chain:AgentExecutor] Agent selected action: {
  "tool": "list_accounts",
  "toolInput": {
    "query": {
      "main": "main"
    },
    "operation_type": "count"
  },
  "log": ""
}
[tool/start] [1:chain:AgentExecutor > 3:tool:DynamicStructuredTool] Entering Tool run with input: "{"query":{"_id":"","referred_by":"","email":"","name":"","owner.name":"","main":"main","created_at":""},"sort":"","operation_type":"count","select":""}"
[tool/end] [1:chain:AgentExecutor > 3:tool:DynamicStructuredTool] [1ms] Exiting Tool run with output: "99"
```

You can see the error thrown here
```
Error in handler Handler, handleAgentAction: Error: need_confirmation
```
But it just continues on to invoke the tool.

Not sure how I would go about a long running function that collects feed back. Seems like a pretty complex solution for this basically simple use case. 

Any other ways you can think off to forcefully exit the chain based on some condition?

**Comment by ppramesi** [comment]
So I've looked into this, and the reason why your original code stops when using gpt-3.5 is because it got stopped by uncaught zod parser error (i.e. sometimes gpt-3 outputs tool parameter without query, which throws error), while gpt-4 handles it correctly. So the current default behavior is actually the chain continuing after it encounters error from inside callback handlers. So if you try to run the following:

```
import { ChatOpenAI } from "langchain/chat_models/openai";
import { initializeAgentExecutorWithOptions } from "langchain/agents";
import { Calculator } from "langchain/tools/calculator";

const runFunc = async function(){
  const controller = new AbortController();
  const tools = [
    new Calculator()
  ];
  const model = new ChatOpenAI({
    modelName: "gpt-3.5-turbo",
    temperature: 0,
    streaming: true,
  });
  try {
    const executor = await initializeAgentExecutorWithOptions(
      tools,
      model,
      {
        verbose: true,
        agentType: "openai-functions",
        returnIntermediateSteps: true,
        maxIterations: 3,
        callbacks: [{
          handleAgentAction(action) {
            controller.abort({
              reason: 'need_confirmation',
              action
            });
            throw new Error("Weeeeeeee")
          },
        }]
      }
    );
    
    const result = await executor.call({
      input: "What 10 * 30?",
      signal: controller.signal
    });
    
    // AbortError
    console.log(result);
  } catch (error) {
    console.log("inside catch", error)
  }  
}

runFunc().then(() => {
  setTimeout(() => {
    console.log("done")
    process.exit(0)
  }, 2000)
}).catch(() => {
  setTimeout(() => {
    console.log("error")
    process.exit(1)
  }, 2000)
})
```

You'll see that it will continue on with the tool call and even the next LLM call. However, if we were to add `throw err` in the catch blocks in the callback manager class's handle function, it will stop the execution right then and there.

FYI the output log for the above code:

```
[chain/start] [1:chain:AgentExecutor] Entering Chain run with input: {
  "input": "What 10 * 30?",
  "signal": {},
  "chat_history": []
}
[llm/start] [1:chain:AgentExecutor > 2:llm:ChatOpenAI] Entering LLM run with input: {
  "messages": [
    [
      {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "schema",
          "SystemMessage"
        ],
        "kwargs": {
          "content": "You are a helpful AI assistant.",
          "additional_kwargs": {}
        }
      },
      {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "schema",
          "HumanMessage"
        ],
        "kwargs": {
          "content": "What 10 * 30?",
          "additional_kwargs": {}
        }
      }
    ]
  ]
}
[llm/end] [1:chain:AgentExecutor > 2:llm:ChatOpenAI] [1.51s] Exiting LLM run with output: {
  "generations": [
    [
      {
        "text": "",
        "message": {
          "lc": 1,
          "type": "constructor",
          "id": [
            "langchain",
            "schema",
            "AIMessage"
          ],
          "kwargs": {
            "content": "",
            "additional_kwargs": {
              "function_call": {
                "name": "calculator",
                "arguments": "{\n  \"input\": \"10 * 30\"\n}"
              }
            }
          }
        }
      }
    ]
  ],
  "llmOutput": {
    "tokenUsage": {}
  }
}
Error in handler Handler, handleAgentAction: Error: Weeeeeeee
[agent/action] [1:chain:AgentExecutor] Agent selected action: {
  "tool": "calculator",
  "toolInput": {
    "input": "10 * 30"
  },
  "log": ""
}
[chain/error] [1:chain:AgentExecutor] [1.53s] Chain run errored with error: "AbortError"
inside catch Error: AbortError
    at EventTarget.<anonymous> (/home/ppramesi/ppramesi-langchains/embeddings-callback/langchain/dist/chains/base.js:92:36)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:737:20)
    at EventTarget.dispatchEvent (node:internal/event_target:679:26)
    at abortSignal (node:internal/abort_controller:314:10)
    at AbortController.abort (node:internal/abort_controller:344:5)
    at Handler.handleAgentAction (/home/ppramesi/ppramesi-langchains/embeddings-callback/examples/src/agents/agent_callbacks.ts:26:24)
    at file:///home/ppramesi/ppramesi-langchains/embeddings-callback/langchain/dist/callbacks/manager.js:199:54
    at consumeCallback (file:///home/ppramesi/ppramesi-langchains/embeddings-callback/langchain/dist/callbacks/promises.js:17:15)
    at file:///home/ppramesi/ppramesi-langchains/embeddings-callback/langchain/dist/callbacks/manager.js:196:58
    at Array.map (<anonymous>)
[tool/start] [1:tool:Calculator] Entering Tool run with input: "10 * 30"
[tool/end] [1:tool:Calculator] [4ms] Exiting Tool run with output: "300"
[llm/start] [1:llm:ChatOpenAI] Entering LLM run with input: {
  "messages": [
    [
      {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "schema",
          "SystemMessage"
        ],
        "kwargs": {
          "content": "You are a helpful AI assistant.",
          "additional_kwargs": {}
        }
      },
      {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "schema",
          "HumanMessage"
        ],
        "kwargs": {
          "content": "What 10 * 30?",
          "additional_kwargs": {}
        }
      },
      {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "schema",
          "AIMessage"
        ],
        "kwargs": {
          "content": "",
          "additional_kwargs": {
            "function_call": {
              "name": "calculator",
              "arguments": "{\"input\":\"10 * 30\"}"
            }
          }
        }
      },
      {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "schema",
          "FunctionMessage"
        ],
        "kwargs": {
          "content": "300",
          "name": "calculator",
          "additional_kwargs": {}
        }
      }
    ]
  ]
}
[llm/error] [1:llm:ChatOpenAI] [2ms] LLM run errored with error: "Cancel: canceled"
done
```

**Comment by dejecj** [comment]
> I'm not sure if this is intended, as it make sense for the callback to not throw an error when it's running in the background (i.e. when LANGCHAIN_CALLBACKS_BACKGROUND is true), but if it's not running in the background, I think it make sense for callbacks to be able to stop execution.

Maybe it can be a config option. For example LANGCHAIN_CALLBACKS_EXIT_ON_ERROR?

**Comment by jacoblee93** [comment]
The blocking behavior is actually for serverless environments where you want to minimize unresolved promises on function exit - however, they will be `await`ed. 

Thanks @ppramesi for the above comment - the callback systems were really designed with tracing/debugging in mind at the moment, so they don't stop execution on errors to allow functionality to continue even if the tracing server is down.

If you await rather than throwing an error, it should do what you're thinking of, but yeah unfortunately I don't think there's a way to completely stop execution here - we could add an env var or maybe return `false` or an error object instead? I know there's also a `HumanTool` concept but I don't think it would fully solve the issue here.

Let me tag @nfcampos who's done more with this (he's very busy at the moment though!).

**Comment by dejecj** [comment]
@jacoblee93 I suppose it makes sense the way it behaves if the callbacks were meant for tracing and logging, etc. You wouldn't want those activities to interrupt the main functions of the app. 

It would be nice to leverage those callbacks for things like this though. Since they are conveniently placed all throughout the execution flow and could easily act as inline hooks for extending the functionality of the chain it self rather than just tracing and logging.

Curious to see what other people think about that.

**Comment by ppramesi** [comment]
imo I think it's not a bad idea to implement the ability to stop chain execution altogether from callback handlers (e.g. when checking if things are going wrong and then stopping execution from inside the handlers). Though I think it'd make more sense to add a property into the base callback handler class that checks whether it should throw an error instead of having it be set by global variable (e.g. through env variable), since if it's set in the global variable, things that shouldn't throw error if failing (e.g. logging and tracing) will also stops execution.

**Comment by stewones** [comment]
same here. all i need is a way to stop the LLM once a tool has been used, is there a way to do that?

here's my log. I wanted to stop the LLM on the `tool/end` stage

```
[chain/start] [1:chain:AgentExecutor] Entering Chain run with input: {
  "input": "talk to a human",
  "signal": {},
  "timeout": 120000,
  "chat_history": [
    {
      "lc": 1,
      "type": "constructor",
      "id": [
        "langchain",
        "schema",
        "SystemMessage"
      ],
      "kwargs": {
        "content": "",
        "additional_kwargs": {}
      }
    }
  ]
}
[llm/start] [1:chain:AgentExecutor > 2:llm:ChatOpenAI] Entering LLM run with input: {
  "messages": [
    [
      {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "schema",
          "SystemMessage"
        ],
        "kwargs": {
          "content": "You are Alicia, a helpful customer support assistant to help answer questions.",
          "additional_kwargs": {}
        }
      },
      {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "schema",
          "SystemMessage"
        ],
        "kwargs": {
          "content": "",
          "additional_kwargs": {}
        }
      },
      {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "schema",
          "HumanMessage"
        ],
        "kwargs": {
          "content": "talk to a human",
          "additional_kwargs": {}
        }
      }
    ]
  ]
}
[llm/end] [1:chain:AgentExecutor > 2:llm:ChatOpenAI] [1.18s] Exiting LLM run with output: {
  "generations": [
    [
      {
        "text": "",
        "generationInfo": {
          "prompt": 0,
          "completion": 0
        },
        "message": {
          "lc": 1,
          "type": "constructor",
          "id": [
            "langchain",
            "schema",
            "AIMessageChunk"
          ],
          "kwargs": {
            "content": "",
            "additional_kwargs": {
              "function_call": {
                "name": "escalate-support",
                "arguments": "{}"
              }
            }
          }
        }
      }
    ]
  ]
}
[agent/action] [1:chain:AgentExecutor] Agent selected action: {
  "tool": "escalate-support",
  "toolInput": {},
  "log": "Invoking \"escalate-support\" with {}\n",
  "messageLog": [
    {
      "lc": 1,
      "type": "constructor",
      "id": [
        "langchain",
        "schema",
        "AIMessageChunk"
      ],
      "kwargs": {
        "content": "",
        "additional_kwargs": {
          "function_call": {
            "name": "escalate-support",
            "arguments": "{}"
          }
        }
      }
    }
  ]
}
[tool/start] [1:chain:AgentExecutor > 3:tool:DynamicStructuredTool] Entering Tool run with input: "{}"
[tool/end] [1:chain:AgentExecutor > 3:tool:DynamicStructuredTool] [1ms] Exiting Tool run with output: "Please provide your name so we can escalate your support request."
[llm/start] [1:chain:AgentExecutor > 4:llm:ChatOpenAI] Entering LLM run with input: {
  "messages": [
    [
      {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "schema",
          "SystemMessage"
        ],
        "kwargs": {
          "content": "You are Alicia, a helpful customer support assistant to help answer questions.",
          "additional_kwargs": {}
        }
      },
      {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "schema",
          "SystemMessage"
        ],
        "kwargs": {
          "content": "",
          "additional_kwargs": {}
        }
      },
      {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "schema",
          "HumanMessage"
        ],
        "kwargs": {
          "content": "talk to a human",
          "additional_kwargs": {}
        }
      },
      {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "schema",
          "AIMessageChunk"
        ],
        "kwargs": {
          "content": "",
          "additional_kwargs": {
            "function_call": {
              "name": "escalate-support",
              "arguments": "{}"
            }
          }
        }
      },
      {
        "lc": 1,
        "type": "constructor",
        "id": [
          "langchain",
          "schema",
          "FunctionMessage"
        ],
        "kwargs": {
          "content": "Please provide your name so we can escalate your support request.",
          "name": "escalate-support",
          "additional_kwargs": {}
        }
      }
    ]
  ]
}
[llm/end] [1:chain:AgentExecutor > 4:llm:ChatOpenAI] [1.44s] Exiting LLM run with output: {
  "generations": [
    [
      {
        "text": "Hello! I'm Alicia, a virtual assistant. How can I help you today? If you have any questions or need assistance with anything, feel free to ask.",
        "generationInfo": {
          "prompt": 0,
          "completion": 0
        },
        "message": {
          "lc": 1,
          "type": "constructor",
          "id": [
            "langchain",
            "schema",
            "AIMessageChunk"
          ],
          "kwargs": {
            "content": "Hello! I'm Alicia, a virtual assistant.",
            "additional_kwargs": {}
          }
        }
      }
    ]
  ]
}
[chain/end] [1:chain:AgentExecutor] [9.99s] Exiting Chain run with output: {
  "output": "Hello! I'm Alicia, a virtual assistant. How can I help you today? If you have any questions or need assistance with anything, feel free to ask."
}
```

**Comment by tutuxxx** [comment]
createOpenAIToolsAgent and createOpenAIFunctionsAgent have the same problem

**Comment by jacoblee93** [comment]
I think we should have some built in abort mechanism for runnables
